#  <p style="text-align: center;">Практическое задание к Семинару 5.<br><br>Урок 5. Тестирование гипотез.</p>

In [31]:
import scipy.stats as stats
import statistics as st
import numpy as np
import math

**АЛГОРИТМ МЕТОДИКИ ТЕСТИРОВАНИЯ  ГИПОТЕЗ**
<div class="alert alert-block alert-warning"><font size=4 color=black><b>1 этап - Формулирование нулевой  $H_{0}$ и  альтернативной гипотез $H_{1}$;<br>
2 этап - Выбор уровня статистической значимости α;<br>
3 этап - Выбор статистического критерия;<br>
4 этап - Расчет наблюдаемого критерия;<br>
5 этап - Сравнение табличного и наблюдаемого значений критерия;<br>
6 этап - Вывод.</b></font></div>

**Z-критерий**
<div class="alert alert-block alert-warning"><font size=4 color=black><b>$z_н=\frac{\bar X-\mu_0}{\sigma/\sqrt n}$ (1)</b></font></div>

**Критерий Стьюдента (t)**
<div class="alert alert-block alert-warning"><font size=4 color=black><b>$t_н=\frac{\bar X-\mu_0}{\sigma_н\sqrt n}$ (2)</b></font></div>


Таблица 1. Z Таблица (положительные значения)

![Positive Z Table](https://datascience.eu/wp-content/uploads/2020/05/image1.png)

Таблица 2. Z Таблица (отрицательные значения)

![Negative Z Table](https://www.ztable.net/wp-content/uploads/2018/11/negativeztable.png)

Таблица 3. Таблица Стьюдента    

![Student Table](https://fsd.multiurok.ru/viewImage.php?image=http://nice-diplom.ru/templates/blue/images/img/form_69.gif)

## ЗАДАНИЕ 1. 
**Когда используется критерий Стьюдента, а когда Z –критерий?**

**<p style="text-align: center;">
    <font size=10 color=#32CD32><b>-------------------------------</b></font><br>
    <font size=4.5 color=#3ac444>РЕШЕНИЕ:</font><br>
<font size=10 color=#32CD32><b>-------------------------------</b></font></p>**

<div class="alert alert-block alert-success"><font size=3 color='b'><b>ОТВЕТ: </b><br>Критерии Стьюдента и Z-критерий, являются параметрическими критериями и используются для статистической проверки гипотезы путем сравнения двух выборочных средних. Оба критерия предполагают, что выборка взята из нормально распределенной генеральной совокупности, и имеется равенство дисперсий в группах. Если известна сигма генеральной совокупности более предпочтительнее ипспользовать Z- критерий, в противном случае -  критерий Стьюдента t. </font></div>

---

## ЗАДАНИЕ 2. 
**Проведите тест гипотезы. Утверждается, что шарики для подшипников, изготовленные автоматическим станком, имеют средний диаметр 17 мм. Используя односторонний критерий с α=0,05, проверить эту гипотезу, если в выборке из n=100 шариков средний диаметр оказался равным 17.5 мм, а дисперсия известна и равна 4 кв. мм.**

**<p style="text-align: center;">
    <font size=10 color=#32CD32><b>-------------------------------</b></font><br>
    <font size=4.5 color=#3ac444>РЕШЕНИЕ:</font><br>
<font size=10 color=#32CD32><b>-------------------------------</b></font></p>**

**1 этап - Формулирование нулевой  $H_{0}$ и  альтернативной гипотез $H_{1}$.**<br>
***$H_{0}$ гипотеза*** - Средний диаметр шариков для подшипников, изготовленные автоматическим станком <b>НЕ ИМЕЕТ</b> различий с диаметром шиариков из выборки. $H_{0}$: $M(X) = 17 => \mu=\mu_0$<br>
***$H_{1}$ гипотеза*** - Средний диаметр шариков для подшипников, изготовленные автоматическим станком <b>ИМЕЕТ</b> различие с диаметром шиариков из выборки. $H_{1}$: $M(X) = 17,5 => \mu>\mu_0$<br><br>
**2 этап - Выбор уровня статистической значимости α.**<br>
Из условия **α = 0,05**.<br><br>
**3 этап - Выбор статистического критерия.**<br>
Т.к. из условия нам известна дисперсия, то будем использовать Z-критерий.<br><br>
**4 этап - Расчет наблюдаемого критерия.**<br>
***Дано:***<br>
$\alpha = 0.05$<br>
$D(X) = 4$<br>
$\sigma=\sqrt 4=2$<br>
n = 100 размер выборки<br>
$\bar X = \mu=17.5$<br>
$\mu_{0}=17$<br>
По формуле (1), находим наблюдаемое значение Z-критерия.

In [32]:
z=(17.5-17)/0.2
print(f'Наблюдаемое значение z-критерия: {z: .4f}')

Наблюдаемое значение z-критерия:  2.5000


**5 этап - Сравнение табличного и наблюдаемого значений критерия.**<br>
Табличное значение z-критерия. Т.к. α = 5% (0.05) и $\mu>\mu_0$, то альтернативная гипотеза находиться справа, а слева лежат 0.95 или 95% всех значений стандартного нормального распределения. По таблице Z-значений сначала ищем долю 0.95. А потом смотрим, на пересечении каких составных частей Z лежит эта доля. $Z_{tab}\approx1.645$.<br><br>

In [33]:
# Проверка правилности определения табличного значения
z_tab=stats.norm.ppf(0.95)
print(f'Табличное значение z-критерия: Z_tab {z_tab: .4f}')

Табличное значение z-критерия: Z_tab  1.6449


$Z_{nab} > Z_{tab} = > 2.5000 > 1.645 $ 

Т.к. $Z_{nab} > Z_{tab}$ отвергаем нулевую гипотезу на уровне значимости  𝛼=0.05 и можем констатировать, что средний диаметр шариков больше 17.

<div class="alert alert-block alert-success"><font size=3 color='b'><b>ОТВЕТ: </b><br> $Z_{nab} > Z_{tab}$ отвергаем нулевую гипотезу на уровне значимости  𝛼=0.05 и можем констатировать, что средний диаметр шариков больше 17.  </font></div>

---

## ЗАДАНИЕ 3. 
**Проведите тест гипотезы. Продавец утверждает, что средний вес пачки печенья составляет 200 г. Из партии извлечена выборка из 10 пачек. Вес каждой пачки составляет: 202, 203, 199, 197, 195, 201, 200, 204, 194, 190. Известно, что их веса распределены нормально. Верно ли утверждение продавца, если учитывать, что доверительная вероятность равна 99%? (Провести двусторонний тест.)**

**<p style="text-align: center;">
    <font size=10 color=#32CD32><b>-------------------------------</b></font><br>
    <font size=4.5 color=#3ac444>РЕШЕНИЕ:</font><br>
<font size=10 color=#32CD32><b>-------------------------------</b></font></p>**

**1 этап - Формулирование нулевой  $H_{0}$ и  альтернативной гипотез $H_{1}$.**<br>
***$H_{0}$ гипотеза*** - Средний вес пачки печенья составляет 200 г. и <b>НЕ ИМЕЕТ</b> различий со средним весом из выборки. $H_{0}$: $M(X) = 200 => \mu=\mu_0$<br>
***$H_{1}$ гипотеза*** - Средний вес пачки печенья составляет 200 г. и <b>ИМЕЕТ</b> различия со средним весом из выборки. $H_{1}: \mu\neq\mu_0$<br><br>
**2 этап - Выбор уровня статистической значимости α.**<br>
Из условия **$\alpha=1-0.99=0.01$**.<br><br>
**3 этап - Выбор статистического критерия.**<br>
Т.к. из условия нам не известны ни среднее, ни дисперсия генеральной совокупности, то будем использовать критерий Стьюдента (t).<br><br>
**4 этап - Расчет наблюдаемого критерия.**<br>
***Дано:***<br>
$\alpha = 0.05$<br>
pechenie = [202, 203, 199, 197, 195, 201, 200, 204, 194, 190]<br>
n = 10 пач. размер выборки<br>
$\mu_{0}=200$<br>
Доверительная вероятность $= 0.99$<br>
Число степеней свободы = $n - 1 = 9$<br>
По формуле (2), находим наблюдаемое значение критерий Стьюдента (t).

In [34]:
pechenie = [202, 203, 199, 197, 195, 201, 200, 204, 194, 190]
st.mean(pechenie)   # sum(f)/len(f)

198.5

In [35]:
t_n=(st.mean(pechenie)-200)/(np.std(pechenie,ddof=1)/len(pechenie)**0.5)
print(f'Наблюдаемое значение критерия Стьюдента (t): {t_n: .4f}')

Наблюдаемое значение критерия Стьюдента (t): -1.0651


**5 этап - Сравнение табличного и наблюдаемого значений критерия.**<br>
По таблице 2 Стьюдента, с учетом $t_t$, для $\alpha/2=0.01/2=0.005$, Доверительная вероятность  = 0.99, Число степеней свободы =  𝑛−1=9, находим значение:<br>
$t_1= 3.25$   <br>
$t_2= -3.25$ <br>
$\Omega_\alpha = (-\infty, -3.25) \cup (3.25, \infty).$

In [36]:
# Проверим
t_t=stats.t.ppf(0.995,len(pechenie)-1)
print(f'Табличное значение критерия Стьюдента (t): {t_t: .4f}')

Табличное значение критерия Стьюдента (t):  3.2498


Сравним полученные значения (поскольку у нас двусторонний критерий, то используем абсолютное значение полученного критерия Стьюдента):

In [37]:
np.abs(t_n)<t_t

True

Т.к. значение $t$ не попадает в критическую область $|t_н|<t_t$, **принимаем** нулевую гипотезу на уровне значимости $\alpha=0.01$

<div class="alert alert-block alert-success"><font size=3 color='b'><b>ОТВЕТ: </b><br> $Z_{nab} > Z_{tab}$ Продавец прав, т.к. значение 𝑡 не попадает в критическую область  $|𝑡_н|<𝑡_𝑡$, принимаем нулевую гипотезу на уровне значимости  𝛼=0.01.</font></div>

---

## ЗАДАНИЕ 4. 
**Есть ли статистически значимые различия в росте дочерей? Рост матерей 172, 177, 158, 170, 178,175, 164, 160, 169 Рост взрослых дочерей: 173, 175, 162, 174, 175, 168, 155, 170, 160.**

1 вариант.

In [46]:
mothers = [172, 177, 158, 170, 178, 175, 164, 160, 169, 165]
daughters = [173, 175, 162, 174, 175, 168, 155, 170, 160, 163]

stats.ttest_rel(mothers, daughters)

TtestResult(statistic=0.6648478531431979, pvalue=0.5228168632983574, df=9)

Т.к. $p-value > \alpha (0,05)$, принемаем $H_0$. Статистически значимых различий нет.

2 вариант.

In [42]:
def equalmean(x,y):
    # вычисляем разность выборок
    dif=mothers-daughters
    n=len(dif)
    # вычисляем среднее разности
    mean_dif=dif.mean()
    # вычисляем несмещенную дисперсию разности
    var_dif=dif.var()*n/(n-1)
    # вычисляем t-статистику
    t=mean_dif/math.sqrt(var_dif)*math.sqrt(n)
    # вычисляем р-уровень
    p = 1 - math.fabs(stats.t.cdf(t,n-1) - stats.t.cdf(-t,n-1))
    return t,p

In [45]:
mothers=np.array([172,177,158,170,178,175,164,160,169,165])
daughters=np.array([173,175,162,174,175,168,155,170,160,163])
t,p=equalmean(mothers,daughters)
print("t-статистика критерия = ",t)
print("p-уровень = ",p)
alpha=float(input("Введите уровень значимости "))
if p>alpha:
    print("статистически значимых различий в росте матерей и дочерей нет")
else:
    print("есть статистически значимые различия в росте матерей и дочерей")
print("\n Проверка с помощью встроенной функции")
res=stats.ttest_rel (a=mothers, b=daughters)
print("t-статистика критерия = ",res.statistic)
print("p-уровень = ",res.pvalue)

t-статистика критерия =  0.664847853143198
p-уровень =  0.5228168632983574
Введите уровень значимости 0.05
статистически значимых различий в росте матерей и дочерей нет

 Проверка с помощью встроенной функции
t-статистика критерия =  0.6648478531431979
p-уровень =  0.5228168632983574
